### DISCUSSION TITLE
OpenMeteo-api Self Hosted

### DISCUSSION THREAD

**Discussion by emcomit** [open-meteo/open-meteo#164] [Q&A]
Dear @patrick-zippenfenig ,
firstly thanks for your great job.
I installed OpenMeteo-api on my Ubuntu workstation following your [Readme](https://github.com/open-meteo/open-meteo/blob/main/docs/getting-started.md) . I added the crontab as described [here](https://github.com/open-meteo/open-meteo/blob/main/Sources/App/Commands/CronjobCommand.swift). In the working directory /var/lib/openmeteo-api the subfolder data is now populated, but calling (for example) the api `http://127.0.0.1:8080/v1/forecast?latitude=52.52&longitude=13.41&current_weather=true&hourly=temperature_2m,relativehumidity_2m,windspeed_10m` I see a lot of null data; Is it normal?
Another question; The marine, ski and air-quality APIs are included in the openmeteo-api source to be installed self Hosted or not? Calling (for example)  `http://127.0.0.1:8080/v1/marine?latitude=54.3213&longitude=10.1348&hourly=wave_height,wave_direction,wind_wave_height,wind_wave_peak_period,swell_wave_height,swell_wave_direction,swell_wave_period` I see all null values...why?

**Comment by patrick-zippenfenig** [comment]
Hm. As soon as some model data is available, the API should pick up whatever is available. Which weather models did you download? ECMWF open-data is usually the smallest and easiest.

Yes, all the other APIs can be self-hosted as well. For each API different weather models need to be downloaded. If you run `openmeteo-api` you will see a list of available download options. Documentation can be improved, but for now I focused on integrating all the major datasets. After ensemble models, I will start another round of documentation updates

**Comment by emcomit** [comment]
Hi @patrick-zippenfenig 
I modified my reply, because yesterday night calling the /vi/forecast it showed me all null values, while today it show values, but not for all.
For example calling http://120.0.0.1:8080/v1/forecast?latitude=45.07&longitude=7.69&hourly=temperature_2m,relativehumidity_2m,apparent_temperature,snowfall,snow_depth,weathercode,windspeed_10m,winddirection_10m,windgusts_10m,soil_temperature_0cm,soil_temperature_6cm I see your correct values, but not for weathercode (where on my self hosted instance are all null), soil_temperature_0cm and soil_temperature_6cm; Those values are all null for my instance.
Have you an idea about this?

Another notice is that AirQuality API work fine, but MarineApis not.
For marine apis (for exemple) I see all null values. I've read that the model source for Marine Weather Data is provided by DWD? How can check if the correct model are downloaded? How can I debug it?
Thanks

[UPDATE]
I found the issue. Calling: 
`openmeteo-api download-iconwave gwam`
it goes in error
```
u:~/OpenMeteoSrc/open-meteo/Sources$ openmeteo-api download-iconwave gwam
[ DEBUG ] Starting OmFileManager (App/OmFileManager.swift:30)
[ INFO ] Downloading domain 'gwam' run '2022-12-06T12:00' (App/IconWaveDownload.swift:55)
[ INFO ] Downloading hour 0 (App/IconWaveDownload.swift:81)
[ INFO ] Downloading file http://opendata.dwd.de/weather/maritime/wave_models/gwam/grib/12/swh/GWAM_SWH_2022120612_000.grib2.bz2 (App/Curl.swift:75)
[ WARNING ] cannotOpenFile(filename: "./data/omfile-gwam/HSURF.om", errno: 2, error: "No such file or directory") (Vapor/Application.swift:107)
[ DEBUG ] Application shutting down (Vapor/Application.swift:136)
[ DEBUG ] Shutting down connection pool [ahc-pool-key: http://opendata.dwd.de:80 TLS-hash: 4283607206265483702] (AsyncHTTPClient/HTTPConnectionPool.swift:83)
[ DEBUG ] connection closed [ahc-connection-id: 0, ahc-http-version: http/1.1, ahc-pool-key: http://opendata.dwd.de:80 TLS-hash: 4283607206265483702] (AsyncHTTPClient/HTTPConnectionPool.swift:483)
Swift/ErrorType.swift:200: Fatal error: Error raised at top level: SwiftPFor2D.SwiftPFor2DError.cannotOpenFile(filename: "./data/omfile-gwam/HSURF.om", errno: 2, error: "No such file or directory")
```
Have you idea why?

**Reply by patrick-zippenfenig** [reply]
Ah yes, this is a good start. Some weather variables are not available from all weather models. `weathercode` is currently only used from DWD ICON. This will change however soon, as I want to derive weather codes for other models as well.

For GWAM the issue is that the directory does not exist. I fixed it here: https://github.com/open-meteo/open-meteo/commit/7d2fb4324da20f0c2ae1500062c2706cf8b85af8 You can just manually create the `omfile-gwam` directory
